File: code-external/mapReduce.py
import threading
from collections import defaultdict

def map_reduce(words, num_threads):
    word_counts = defaultdict(int)
    chunk_size = len(words) // num_threads

    def process_chunk(chunk):
        local_counts = defaultdict(int)
        for word in chunk:
            local_counts[word] += 1
        with lock:
            for word, count in local_counts.items():
                word_counts[word] += count

    lock = threading.Lock()
    threads = []
    for i in range(num_threads):
        chunk = words[i * chunk_size: (i + 1) * chunk_size]
        thread = threading.Thread(target=process_chunk, args=(chunk,))
        thread.start()
        threads.append(thread)

    for thread in threads:
        thread.join()

    return dict(word_counts)


if __name__ == "__main__":
    import time
    start_time = time.time()
    with open("citiesOrders.txt", "r") as file:
        words = [line.strip() for line in file]

   
    result = map_reduce(words, 4)
    elapsed_time = time.time() - start_time

    for word, count in result.items():
        print(f"{word}: {count}")

    print(f"Elapsed: {elapsed_time:.2f}s")

File: code-orig/mapReduce.rs
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::thread;
use std::fs;
fn main() {
    let input = fs::read_to_string("citiesOrders.txt")
        .expect("Could not read file");
    
    let words = input.lines()
        .map(|line| line.to_string())
        .collect::<Vec<String>>();
    
    use std::time::Instant;
    let now = Instant::now();
    let result = map_reduce(words, 4);
    let elapsed = now.elapsed();
    for (word, count) in result {
        println!("{}: {}", word, count);
    }
    println!("Elapsed: {:.2?}", elapsed);
}

fn map_reduce(words: Vec<String>, num_threads: usize) -> HashMap<String, u32> {
    let word_counts: Arc<Mutex<HashMap<String, u32>>> = Arc::new(Mutex::new(HashMap::new()));
let mut handles = vec![];

for chunk in words.chunks(words.len() / num_threads) {
    let word_counts = Arc::clone(&word_counts);
    let chunk = chunk.to_vec();
    let handle = thread::spawn(move || {
        // let mut local_counts = HashMap::new();
        let mut local_counts = HashMap::with_capacity(chunk.len());
        for word in &chunk {
            *local_counts.entry(word.to_string()).or_insert(0) += 1;
        }
        let mut word_counts = word_counts.lock().unwrap();
        for (word, count) in local_counts {
            *word_counts.entry(word).or_insert(0) += count;
        }
    });
    handles.push(handle);
}

for handle in handles {
    handle.join().unwrap();
}

Arc::try_unwrap(word_counts).unwrap().into_inner().unwrap()
}
